src/i_winmusic.c

```diff
@@ -1660,28 +1663,13 @@ static void I_WIN_ShutdownMusic(void)
     {
         MidiError("midiStreamRestart", mmr);
     }
-    WaitForSingleObject(hBufferReturnEvent, INFINITE);
+    WaitForSingleObject(hBufferReturnEvent, WAIT_TIME);
     mmr = midiStreamStop(hMidiStream);
     if (mmr != MMSYSERR_NOERROR)
     {
         MidiError("midiStreamStop", mmr);
     }
 
-    if (buffer.data)
```

Comment:
Unfortunately, removing this section breaks MIDI device switching and freezes the game. I think `midiOutUnprepareHeader()` has to be called no matter what before closing the stream.
```
midiStreamOut: The current MIDI Mapper setup refers to a MIDI device that is not installed on the system.  Use MIDI Mapper to edit the setup..
```

Reply:
This https://github.com/fabiangreffrath/woof/pull/1109/commits/32ea5e3c25ddebc7a4d6f64fe6213b0dae715697 works stable for me with ASan enabled. After the process exits, all memory is freed anyway, so I think the 8K buffer is not a big deal.

Reply:
Yes, that works for me now. Reusing the buffer makes sense.

Reply:
Interestingly, I always get the same `hMidiStream` handle when I close/open the stream. Perhaps we should "prepare" the buffer again if the handle changes.

Reply:
Yes, that's important. Should the device only be partially shutting down unless it's at exit?
```c
static void I_WIN_ShutdownMusic(boolean at_exit)
{
    // ...
    buffer.position = 0;

    if (!at_exit)
    {
        return;
    }

    mmr = midiStreamClose(hMidiStream);
    // ...
}
```
Or maybe a new function pointer, e.g. `midi_player_module->I_StopDevice()` called by `I_SetMidiPlayer()`. Other modules can just use the same shutdown function.
```c
music_module_t music_win_module =
{
    I_WIN_ShutdownMusic,
    I_WIN_StopDevice,
}

music_module_t music_xxx_module =
{
    I_xxx_ShutdownMusic,
    I_xxx_ShutdownMusic, // No StopDevice function for this module.
}

Reply:
This works for me on Win10 and Win7: https://github.com/fabiangreffrath/woof/pull/1109/commits/995d0779a72170afe18769002ef553e9b59ff288

Reply:
"Preparing a header that has already been prepared has no effect, and the function returns MMSYSERR_NOERROR." from https://learn.microsoft.com/en-us/windows/win32/api/mmeapi/nf-mmeapi-midioutprepareheader

Reply:
> Do you think the buffer size 8192 is enough?

I investigated further and that should be plenty: [woof_winmidi.log](https://github.com/fabiangreffrath/woof/files/11782603/woof_winmidi.log)

The largest burst of events in a single `WriteBuffer()` call is the reset process, as expected. I don't think MIDIs will ever come close because only one SysEx event is sent at a time and only 4 standard events are ever sent simultaneously.

Reply:
Some hard numbers for reference:

Each MIDI event is 12 bytes (3 x DWORD), so with a buffer of 8192 bytes that's 682 events that can be sent simultaneously. There's only two ways to reach that limit:

1. We send a very large set of reset events without thinking.
2. The MIDI file contains a malicious SysEx message.

Let's look at (1):

Reset process:
- 384 bytes: `StopSound()`
- 12 + 8 bytes: "GM System On" (up to 1344 bytes if `winmm_reset_type` is "None")
- 1152 bytes: `ResetPitchBendSensitivity()`
- 192 bytes: `ResetVolume()`
- 12 bytes: `SendDelayMsg()` if `winmm_reset_delay` > 0

Total:
- Default: 1748 bytes (144 events + 1 SysEx message)
- Worst case: 4832 bytes (257 events)

Therefore, the buffer is large enough for (1). Now let's look at (2):

In theory, there could be a MIDI file with a SysEx message greater than 8180 bytes. The MIDI file spec allows up to max variable length quantity 0x0FFFFFFF which fits in a DWORD, but no one is going to put a 268 MB MIDI in a WAD. For comparison, loopMIDI only handles SysEx messages up to 512 kB. Regardless, maybe we should ignore SysEx messages over a certain length to avoid reallocating/`midiOutUnprepareHeader()`. I suggest something reasonable like 1024 bytes.

Reply:
Thanks for the numbers!

> Regardless, maybe we should ignore SysEx messages over a certain length to avoid reallocating/`midiOutUnprepareHeader()`. I suggest something reasonable like 1024 bytes.

I think we can allow `midiOutUnprepareHeader()` in some rare cases, most of the time it works. I've caught the crash a few times but have never been able to reproduce it.

Reply:
Okay, sounds good to me.